### PR TITLE
Add test for mutually recursive remote functions.

### PR DESCRIPTION
**Pull request by robertnishihara**
We can do this now that we export remote functions lazily.

Now that we export remote functions lazily, we don't seem to need the cloudpickle workaround anymore.

### CODE REVIEW

**Comment by mehrdadn** [comment]
(@453)

**Comment by robertnishihara** [comment]
As a note to myself, we may be able to move `test_recursion.py` inside of a normal pytest module in this PR.

**Comment by mehrdadn** [comment]
Could you refresh my memory what the issue was? I've tried all sorts of examples and can't find a single one whose behavior changed with the workaround. I've tried older versions of CloudPickle as far as 0.5.3 (two years old).

File: python/ray/tests/test_basic.py [hunk 1]
```diff
@@ -319,6 +319,38 @@ def h():
...
+        return n * ray.get(factorial.remote(n - 1))
+
+    assert ray.get(factorial.remote(0)) == 1
+    assert ray.get(factorial.remote(1)) == 1
+    assert ray.get(factorial.remote(2)) == 2
+    assert ray.get(factorial.remote(3)) == 6
+    assert ray.get(factorial.remote(4)) == 24
+    assert ray.get(factorial.remote(5)) == 120
+
+    # Test remote functions that both call each other.
```
Comment (edoakes) [inline python/ray/tests/test_basic.py]:
```suggestion
    # Test remote functions that recursively call each other.
```

File: python/ray/tests/test_basic.py [hunk 2]
```diff
@@ -319,6 +319,38 @@ def h():
 
     assert ray.get(f.remote()) == (1, 2)
 
+    # Test a recursive remote function.
```
Comment (edoakes) [inline python/ray/tests/test_basic.py]:
```suggestion
    # Test a remote function that recursively calls itself.
```

**Review by edoakes — APPROVED** [review]
LGTM

**Pull request merged**